[cxx-interop] Fix SIL deserialization error

egorzhdan opened the pull request:
Make sure nested namespaces are added to the SwiftLookupTable when they are declared in a non-canonical redecl of the parent namespace.

This was blocking C++ interop adoption in SwiftCompilerSources.

egorzhdan committed:
[cxx-interop] Fix SIL deserialization error

Make sure nested namespaces are added to the SwiftLookupTable when they are declared in a non-canonical redecl of the parent namespace.

This was blocking C++ interop adoption in SwiftCompilerSources.

egorzhdan commented:
@swift-ci please smoke test

egorzhdan commented:
This is a requirement for https://github.com/apple/swift/pull/42297.

egorzhdan commented:
@swift-ci please clean smoke test macOS

egorzhdan commented:
Hmm that wasn't a clean build 🤔

egorzhdan commented:
@swift-ci please clean smoke test macOS platform

egorzhdan commented:
@swift-ci please clean test macOS platform

egorzhdan commented:
@swift-ci please clean smoke test

zoecarver commented on test/Interop/Cxx/class/nested-records-module-interface.swift:
```diff
@@ -82,7 +82,7 @@
 // CHECK:   struct ForwardDeclaredFriend {
 // CHECK:     init()
 // CHECK:   }
-// CHECK:   static func takesFriend(_ b: NestedDeclIsAFirstForwardDeclaration.ForwardDeclaredFriend)
```
This seems odd. Any ideas why this changed?

egorzhdan replied:
Previously we went through all the decls in a namespace, and emitted them, and now we request a canonical decl for each of the decls in a namespace, to make sure we don't emit the same decl multiple times if it has redecls inside different redecls of the namespace.
Different redecls of this function have different parameter names, so I had to update this test.

zoecarver reviewed (approved):
🚢 it

egorzhdan commented:
This builds successfully on Linux/Windows CI, but fails on macOS. Locally for me it builds fine on macOS. I'll need to debug this.

egorzhdan commented:
@swift-ci please smoke test macOS platform

egorzhdan commented:
@swift-ci please smoke test macOS platform

egorzhdan commented:
@swift-ci please smoke test macOS platform

egorzhdan commented:
@swift-ci please smoke test macOS platform

egorzhdan committed:
[cxx-interop] Fix AST serialization error

Do not attempt to serialize C++ builtin decls. This caused an assertion failure in `clang::ASTWriter::getDeclID` when building SwiftCompilerSources:

```
Assertion failed: (DeclIDs.find(D) != DeclIDs.end() && "Declaration not emitted!"), function getDeclID, file ASTWriter.cpp, line 5272.
Please submit a bug report (https://swift.org/contributing/#reporting-bugs) and include the project and the crash backtrace.
Stack dump:
0.	<eof> parser at end of file
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  swift-frontend           0x00000001123e21c7 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 39
1  swift-frontend           0x00000001123e10e8 llvm::sys::RunSignalHandlers() + 248
2  swift-frontend           0x00000001123e2810 SignalHandler(int) + 288
3  libsystem_platform.dylib 0x00007fff204a0d7d _sigtramp + 29
4  libsystem_platform.dylib 0x0000000000000008 _sigtramp + 18446603339974439592
5  libsystem_c.dylib        0x00007fff203b0406 abort + 125
6  libsystem_c.dylib        0x00007fff203af7d8 err + 0
7  swift-frontend           0x000000011345b983 clang::ASTWriter::getDeclID(clang::Decl const*) (.cold.1) + 35
8  swift-frontend           0x0000000110a584b2 clang::ASTWriter::getDeclID(clang::Decl const*) + 322
9  swift-frontend           0x000000010e36649c llvm::OnDiskChainedHashTableGenerator<(anonymous namespace)::BaseNameToEntitiesTableWriterInfo>::Emit(llvm::raw_ostream&, (anonymous namespace)::BaseNameToEntitiesTableWriterInfo&) + 1244
10 swift-frontend           0x000000010e364927 swift::SwiftLookupTableWriter::writeExtensionContents(clang::Sema&, llvm::BitstreamWriter&) + 1351
11 swift-frontend           0x0000000110a59b62 clang::ASTWriter::WriteModuleFileExtension(clang::Sema&, clang::ModuleFileExtensionWriter&) + 946

```

egorzhdan commented:
@swift-ci please smoke test

egorzhdan commented:
I added one more commit with a change in `ClangImporter.cpp` to fix the build on macOS, @zoecarver could you please take another look?

zoecarver reviewed (approved):
Send it! Thanks again, Egor.
